clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+     */
+
+    void startReconcile(Assignment assignment) {
+        SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(assignment);
+        SortedSet<TopicPartition> partitionsToAssign = getPartitionsToAssign(assignment);
+
+        if (partitionsToRevoke.isEmpty() && partitionsToAssign.isEmpty()) {
+            log.debug("Skipping invocation of {} callbacks as no partitions changed in the new assignment",
+                    ConsumerRebalanceListener.class.getSimpleName());
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        PartitionReconciliationStartedEvent event = new PartitionReconciliationStartedEvent(
+                partitionsToRevoke,
+                partitionsToAssign
+        );
+        backgroundEventQueue.add(event);
+    }
+
+    /**
+     * This method should be invoked to signal the completion of a successful {@link TopicPartition partition}
+     * assignment reconciliation. Specifically, it is to be executed on background thread <em>after</em> the
+     * {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} and
+     * {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)} callbacks have completed execution on
+     * the application thread. It should update the set of
+     * {@link SubscriptionState#assignedPartitions() assigned partitions} based on the
+     * given partitions.
+     *
+     * @param revokedPartitions Set of {@link TopicPartition partitions} that were revoked
+     * @param assignedPartitions Set of {@link TopicPartition partitions} that were assigned
+     * @see MembershipManager#completeReconcile(Set, Set, Optional)
+     */
+    void completeReconcile(Set<TopicPartition> revokedPartitions, Set<TopicPartition> assignedPartitions) {
+        log.debug("{} callbacks were successfully invoked", ConsumerRebalanceListener.class.getSimpleName());
+
+        Set<TopicPartition> newAssignment = new HashSet<>(subscriptions.assignedPartitions());
+        newAssignment.addAll(assignedPartitions);
+        newAssignment.removeAll(revokedPartitions);
+        subscriptions.assignFromSubscribed(newAssignment);
+    }
+
+    /**
+     * Determine which partitions are newly revoked. This is done by comparing the
+     * {@link Assignment#topicPartitions() target set from the assignment} against the
+     * {@link SubscriptionState#assignedPartitions() current set}. The returned set of
+     * {@link TopicPartition partitions} are composed of any partitions that are in the current set but
+     * are no longer in the target set.
+     *
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to revoke
+     */
+
+    SortedSet<TopicPartition> getPartitionsToRevoke(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
```

Comment:
Is there a reason why we use a SortedSet?

Reply:
Partially due to logging, but yes, it was that way before.